### PR TITLE
add NFT

### PR DESCRIPTION
**Pull request by Ashuaidehao**

### CODE REVIEW

File: csharp/NFT/NFTContract.Helper.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+        private static void StoragePut(byte[] key, byte[] value)
+        {
+            Storage.Put(Storage.CurrentContext, key, (ByteString)value);
+        }
+
+        private static void StoragePut(byte[] key, BigInteger value)
+        {
+            Storage.Put(Storage.CurrentContext, key, value);
+        }
```
Comment (shargon) [inline csharp/NFT/NFTContract.Helper.cs:39]: I think that these methods are not needed

File: csharp/NFT/NFTContract.Helper.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+namespace NFT
+{
+    partial class NFTContract
+    {
+        /// <summary>
+        /// Convert To BigInteger Safely
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private static BigInteger ToBigInteger(ByteString value) => value is null ? 0 : (BigInteger)value;
```
Comment (shargon) [inline csharp/NFT/NFTContract.Helper.cs:19]: Reuse this method? or move this to SDK? https://github.com/neo-project/neo-devpack-dotnet/blob/522a12fcf5dbe6e4d21b1f2c9082bd30d226ddae/src/Neo.SmartContract.Framework/Helper.cs#L33

File: csharp/NFT/NFTContract.cs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+            var tokenBalanceMap = Storage.CurrentContext.CreateMap(CreateStorageKey(Prefix_TokenBalance, owner));
+            var balanceValue = tokenBalanceMap.Get(tokenId);
+            var balance = balanceValue is null ? 0 : (BigInteger)balanceValue;
+            if (balance < amount) return false;
+
+            var totalSupply = TotalSupply();
+            balance -= amount;
+            totalSupply -= amount;
+            tokenBalanceMap.Put(tokenId, balance);
```
Comment (steven1227) [inline csharp/NFT/NFTContract.cs:137]: Here you should also update the tokenOwnerMap and tokenOfMap when the balance == amount?

File: csharp/NFT/NFTContract.cs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+            }
+            ContractManagement.Update(script, manifest);
+            return true;
+        }
+
+        public static bool Destroy()
+        {
+            if (!Runtime.CheckWitness(superAdmin))
+            {
+                return false;
```
Comment (shargon) [inline csharp/NFT/NFTContract.cs:236]: Throw?

File: csharp/NFT/NFTContract.cs [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            // Notify
+            TransferNotify(from, to, amount, tokenId);
+            return true;
+        }
+
+        public static bool Migrate(ByteString script, string manifest)
+        {
+            if (!Runtime.CheckWitness(superAdmin))
+            {
+                return false;
```
Comment (shargon) [inline csharp/NFT/NFTContract.cs:222]: Throw?

File: csharp/NFT/NFTContract.cs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+        public static string Properties(ByteString tokenid)
+        {
+            byte[] key = CreateStorageKey(Prefix_Properties, tokenid);
+            return Storage.Get(Storage.CurrentContext, key);
+        }
+
+        public static bool Mint(ByteString tokenId, UInt160 owner, byte[] properties)
+        {
+            if (!Runtime.CheckWitness(superAdmin)) return false;
```
Comment (shargon) [inline csharp/NFT/NFTContract.cs:95]: Throw?

File: csharp/NFT/NFTContract.cs (hunk 4 above)
Comment (Ashuaidehao) [reply csharp/NFT/NFTContract.cs]: Ok

File: csharp/NFT/NFT.csproj [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFramework>net5.0</TargetFramework>
+    </PropertyGroup>
+    <ItemGroup>
+        <PackageReference Include="Neo3.Compiler.Dev" Version="3.1.0" />
```
[the remark this answers is not in the document]
Comment (Ashuaidehao) [reply csharp/NFT/NFT.csproj:6]: I packed the RC1 version Compiler, preventing users from using the incompatible 
 compiler.

**Pull request closed without merging**